clusterloader2/pkg/measurement/common/cilium_endpoint_propagation_delay.go

```diff
@@ -0,0 +1,160 @@
+		klog.Fatalf("Cannot register %s: %v", cepPropagationDelayMeasurementName, err)
+	}
+}
+
+type cepPropagationDelayGatherer struct{}
+
+// cepPropagationDelayMetricMap contains timestamps in the outer map,
+// and buckets and their sizes in the inner map.
+type cepPropagationDelayMetricMap map[string]map[string]int
+
+func (c *cepPropagationDelayGatherer) Gather(executor QueryExecutor, startTime, endTime time.Time, config *measurement.Config) ([]measurement.Summary, error) {
+	cepPropagationDelay, err := c.gatherCepPropagationDelay(executor, endTime)
+	if err != nil {
+		return nil, err
+	}
+
+	content, err := util.PrettyPrintJSON(cepPropagationDelay)
+	if err != nil {
+		return nil, err
+	}
+
+	summaries := []measurement.Summary{measurement.CreateSummary(cepPropagationDelayMeasurementName, "json", content)}
+	return summaries, validateCEPPropagationDelay(cepPropagationDelay, config)
+}
+
+func (c *cepPropagationDelayGatherer) gatherCepPropagationDelay(executor QueryExecutor, endTime time.Time) (cepPropagationDelayMetricMap, error) {
+	samples, err := executor.Query(cepPropagationDelayQuery, endTime)
+	if err != nil {
+		return nil, err
+	}
+
+	// Query the data on fixed intervals to get accurate data
+	// from multiple snapshots. Stop when the query returns empty.
+	queryTime := endTime
+	for {
+		queryTime = queryTime.Add(-queryInterval)
+		if queryTime.IsZero() {
+			break
+		}
+		newSamples, err := executor.Query(cepPropagationDelayQuery, queryTime)
+		if err != nil {
+			klog.V(2).Infof("Got error querying Prometheus: %v", err)
+			break
+		}
+		if len(newSamples) == 0 {
+			break
+		}
+		samples = append(samples, newSamples...)
+	}
+
+	extractSampleData := func(sample *model.Sample) (string, string, int) {
+		return sample.Timestamp.String(), string(sample.Metric["le"]), int(math.Round(float64(sample.Value)))
+	}
+
+	result := make(cepPropagationDelayMetricMap)
+	for _, sample := range samples {
+		timestamp, bucket, value := extractSampleData(sample)
+		if _, ok := result[timestamp]; !ok {
+			result[timestamp] = make(map[string]int)
+		}
+		result[timestamp][bucket] = value
+	}
+	return result, nil
+}
+
+func validateCEPPropagationDelay(result cepPropagationDelayMetricMap, config *measurement.Config) error {
+	bucketSLO, err := util.GetStringOrDefault(config.Params, "bucketSLO", defaultBucketSLO)
+	if err != nil || len(bucketSLO) == 0 {
+		klog.V(2).Infof("Using defaultBucketSLO: %s, because bucketSLO param is invalid: %v", defaultBucketSLO, err)
+		bucketSLO = defaultBucketSLO
+	}
+
+	percentileSLO, err := util.GetFloat64OrDefault(config.Params, "percentileSLO", defaultPercentileSLO)
+	if err != nil || percentileSLO == 0 {
+		klog.V(2).Infof("Using defaultPercentileSLO: %f, because percentileSLO param is invalid: %v", percentileSLO, err)
+		percentileSLO = defaultPercentileSLO
+	}
+
+	for _, buckets := range result {
+		totalEvents := buckets[bucketAllEntries]
+		acceptedDelayEvents := buckets[bucketSLO]
```

Comment:
This is slightly brittle - if we provide a `bucketSLO` argument in the measurement's params that isn't defined by the metric, then this will always evaluate to `0` and the measurement will always fail.

OTOH, I believe there is no good way of statically ensuring that we provide a correct `bucketSLO` value. Could we at least point to the definition of the metric in a comment above the definition of `cepPropagationDelayQuery` variable?

Reply:
I agree on both points. I added the link to the file where the metric and its buckets are defined.

Reply:
Done.

Reply:
Thanks. Done.